### PR TITLE
Adds a separate option to preallocate nvjPEG2k memory

### PR DESCRIPTION
**Pull request by JanuszL**
- decouples nvJPEG and nvJPEG2k options to preallocate decoder memory as usually user will have a pure JPEG or pure JPEG2k dataset and doesn't want to waste memory for both decoders

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It decouples nvJPEG and nvJPEG2k options to preallocate decoder memory as usually user will have a pure JPEG or pure JPEG2k dataset and doesn't want to waste memory for both decoders

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     decouples nvJPEG and nvJPEG2k options to preallocate decoder memory as usually user will have a pure JPEG or pure JPEG2k dataset and doesn't want to waste memory for both decoders
 - Affected modules and functionalities:
     image decoder
 - Key points relevant for the review:
     NA
 - Validation and testing:
     CI
 - Documentation (including examples):
     NA

**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_decoder_decoupled_api.h [hunk 1]
```diff
@@ -158,18 +160,21 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
...
+                               device_memory_padding_jpeg2k);
+      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::GPU,
+                               device_memory_padding_jpeg2k);
+    }
+    if (host_memory_padding_jpeg2k > 0) {
+      nvjpeg_memory::AddBuffer(nvjpeg2k_thread_id, kernels::AllocType::Pinned,
+                               host_memory_padding_jpeg2k);
     }
     nvjpeg2k_thread_.AddWork([this, device_memory_padding, host_memory_padding](int) {
       nvjpeg2k_handle_ = NvJPEG2KHandle(device_memory_padding > 0 ? &nvjpeg2k_dev_alloc_ : nullptr,
```
Comment (banasraf) [inline dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_decoder_decoupled_api.h:180]: `device_memory_padding_jpeg2k` and `host_memory_padding_jpeg2k`

Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_decoder_decoupled_api.h]: Done

File: dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_memory.cc [hunk 2]
```diff
@@ -182,7 +193,10 @@ void AddBuffer(std::thread::id thread_id, AllocType alloc_type, size_t size) {
 void DeleteAllBuffers(std::thread::id thread_id) {
   std::shared_lock<std::shared_timed_mutex> lock(buffer_pool_mutex_);
   auto it = buffer_pool_.find(thread_id);
-  assert(it != buffer_pool_.end());
+  // not buffers have been preallocated/returned to the pool
```
Comment (jantonguirao) [inline dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_memory.cc:196]:
```suggestion
  // no buffers have been preallocated/returned to the pool
```

File: dali/operators/decoder/image_decoder.cc [hunk 3]
```diff
@@ -54,6 +54,19 @@ will occur during the pipeline execution. One way to find the ideal value is to
...
 allocation value that was printed in the statistics.)code",
       16*1024*1024)
+  .AddOptionalArg("device_memory_padding_jpeg2k",
+      R"code(Applies **only** to the ``mixed`` backend type.
+
+The padding for nvJPEG2k's device memory allocations, in bytes. This parameter helps to avoid
+reallocation in nvJPEG2k when a larger image is encountered, and the internal buffer needs to be
+reallocated to decode the image.
+
+If a value greater than 0 is provided, the operator preallocates one device buffer of the
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:64]: "one device buffer of the requested size per thread" -> Not true. See allocations below.

File: dali/operators/decoder/image_decoder.cc [hunk 4]
```diff
@@ -67,6 +80,19 @@ allocations will occur during the pipeline execution. One way to find the ideal
...
 the largest allocation value that is printed in the statistics.)code",
       8*1024*1024)  // based on ImageNet heuristics (8MB)
+  .AddOptionalArg("host_memory_padding_jpeg2k",
+      R"code(Applies **only** to the ``mixed`` backend type.
+
+The padding for nvJPEG2k's host memory allocations, in bytes. This parameter helps to prevent
+the reallocation in nvJPEG2k when a larger image is encountered, and the internal buffer needs
+to be reallocated to decode the image.
+
+If a value greater than 0 is provided, the operator preallocates two (because of double-buffering)
```
Comment (jantonguirao) [inline dali/operators/decoder/image_decoder.cc:90]: Again, this is not true.

Comment (JanuszL) [reply dali/operators/decoder/image_decoder.cc:90]: Done

File: dali/operators/decoder/image_decoder.cc (hunk 3 above)
Comment (JanuszL) [reply dali/operators/decoder/image_decoder.cc:64]: Done

File: dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_memory.cc (hunk 2 above)
Comment (JanuszL) [reply dali/operators/decoder/nvjpeg/decoupled_api/nvjpeg_memory.cc:196]: Done

**Comment by JanuszL** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Review by banasraf — APPROVED** [review]

**Pull request merged**